### PR TITLE
Initial SAI port implementation

### PR DESCRIPTION
**Pull request by DanG100**
* dataplane/internal/engine is now a gRPC service
* migrate CreateLocalPort and CreateExternalPort to CreatePort RPC
* extend base class to have both forwarding and dataplane client 
* Moved saving of attributes to base class instead of caller since classs may additional attributes

### CODE REVIEW

File: dataplane/standalone/port.h [hunk 1]
```diff
@@ -28,13 +32,18 @@ extern "C" {
...
+       std::shared_ptr<lemming::dataplane::Dataplane::Stub> dplane)
+      : APIBase(id, mgr, fwd, dplane) {}
   ~Port() = default;
   sai_status_t create(_In_ uint32_t attr_count,
                       _In_ const sai_attribute_t* attr_list);
   sai_status_t set_attribute(_In_ const sai_attribute_t* attr);
+
+ private:
+  static std::unordered_map<std::string, std::vector<int>> parseLaneMap();
+  static std::unordered_map<std::string, std::vector<int>> laneMap;
```
Comment (wenovus) [inline dataplane/standalone/port.h:46]: (no action required) What's a lane map?

File: dataplane/standalone/port.cc [hunk 2]
```diff
@@ -14,13 +14,113 @@
...
+    std::vector<std::string> lanes = absl::StrSplit(elems[1], ",");
+    for (auto lane : lanes) {
+      res[elems[0]].push_back(std::stoi(lane));
+    }
+  }
+  return res;
+}
+
+std::unordered_map<std::string, std::vector<int>> Port::laneMap =
+    Port::parseLaneMap();
```
Comment (wenovus) [inline dataplane/standalone/port.cc:126]: Put this in the header as an in-class initializer?

File: dataplane/standalone/common.cc [hunk 3]
```diff
@@ -156,5 +159,26 @@ sai_object_id_t AttributeManager::entry_to_switch_id(sai_object_type_t type,
     default:
       throw "Invalid type";
   }
-  return swID;
+  return std::to_string(swID);
+}
+
+sai_status_t APIBase::create(common_entry_t id, uint32_t attr_count,
```
Comment (wenovus) [inline dataplane/standalone/common.cc:165]: input id is not being used

File: dataplane/internal/engine/engine.go [hunk 4]
```diff
@@ -80,7 +83,7 @@ func New(ctx context.Context, id string, c fwdpb.ForwardingClient) (*Engine, err
 			},
 		},
 	}
-	if _, err := c.TableCreate(ctx, v4FIB); err != nil {
+	if _, err := e.Engine.TableCreate(ctx, v4FIB); err != nil {
```
Comment (wenovus) [inline dataplane/internal/engine/engine.go:86]: Reading this is a bit confusing: an engine that has an engine. Can add a doc comment for this embedded variable how it's different.

File: dataplane/forwarding/fwdport/ports/kernel.go [hunk 5]
```diff
@@ -172,6 +174,9 @@ func (kernelBuilder) Build(portDesc *fwdpb.PortDesc, ctx *fwdcontext.Context) (f
 			return nil, fmt.Errorf("failed to set sec promisc on: %v", err)
 		}
 	}
+	if err := os.WriteFile(fmt.Sprintf("/proc/sys/net/ipv4/conf/%s/arp_ignore", kp.Kernel.GetDeviceName()), []byte("2"), 0o600); err != nil {
+		return nil, fmt.Errorf("failed to set arp_ignore to 2: %v", err)
+	}
```
Comment (wenovus) [inline dataplane/forwarding/fwdport/ports/kernel.go:179]: This looks like it lost its comment on the move.

File: dataplane/standalone/common.cc (hunk 3 above)
Comment (DanG100) [reply dataplane/standalone/common.cc:165]: the id will be used classes that inherit from this class

File: dataplane/internal/engine/engine.go (hunk 4 above)
Comment (DanG100) [reply dataplane/internal/engine/engine.go]: agreed, renamed forwarding.Engine to forwarding.Server

File: dataplane/forwarding/fwdport/ports/kernel.go (hunk 5 above)
Comment (DanG100) [reply dataplane/forwarding/fwdport/ports/kernel.go:180]: good catch

File: proto/dataplane/dataplane.proto [hunk 6]
```diff
@@ -35,4 +36,20 @@ message Route {
...
   repeated NextHop next_hops = 3;
+}
+
+
+message CreatePortRequest {
+  forwarding.PortType type = 1; 
+  string id = 2;
+  oneof src {
+    string kernel_dev = 3;
+  }
```
[the remark this answers is not in the document]
Comment (DanG100) [reply proto/dataplane/dataplane.proto]: added

File: dataplane/standalone/port.h (hunk 1 above)
Comment (DanG100) [reply dataplane/standalone/port.h:47]: added comment, ports are creates with a list of lanes. the lanemap is map of lanes to veths

File: dataplane/standalone/port.cc (hunk 2 above)
Comment (DanG100) [reply dataplane/standalone/port.cc:126]: in-class initializer are only allows for literals

Comment (wenovus) [reply dataplane/standalone/port.cc:126]: Ok. Then in the initializer list or the constructor?

**Review by wenovus — APPROVED** [review]

File: dataplane/standalone/port.cc (hunk 2 above)
Comment (DanG100) [reply dataplane/standalone/port.cc:126]: this is a static member, it can't be initialized in the constructor

**Pull request merged**